compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java

```diff
@@ -671,22 +672,34 @@ public void visit(BLangWhile whileNode, AnalyzerData data) {
                 data.breakAsLastStatement || data.statementReturnsPanicsOrFails, true,
                 data.potentiallyInvalidAssignmentInLoopsInfo);
 
-        decrementLoopCount(data);
         data.failureHandled = failureHandled;
-        if (data.booleanConstCondition != symTable.trueType || data.breakStmtFound) {
+        if (isUnreachableBlock(whileNode, data)) {
+            data.infiniteLoop = true;
             data.statementReturnsPanicsOrFails = prevStatementReturnsPanicsOrFails;
             data.continueAsLastStatement = prevContinueAsLastStatement;
             data.breakAsLastStatement = prevBreakAsLastStatement;
+        } else if (data.booleanConstCondition != symTable.trueType || data.breakStmtFound) {
+            data.statementReturnsPanicsOrFails = prevStatementReturnsPanicsOrFails;
+            data.continueAsLastStatement = prevContinueAsLastStatement;
+            data.breakAsLastStatement = prevBreakAsLastStatement;
+            data.infiniteLoop = previousInfiniteLoop;
         } else {
             data.statementReturnsPanicsOrFails = true;
         }
+
+        decrementLoopCount(data);
         data.breakStmtFound = prevBreakStmtFound;
 
         analyzeOnFailClause(whileNode.onFailClause, data);
 
         data.loopAndDoClauseEnvs.pop();
     }
 
+    public boolean isUnreachableBlock(BLangWhile whileNode, AnalyzerData data) {
```

Comment:
It would be better to use a name for the function that implies checking for an infinite loop, instead of using `isUnreachableBlock`.